### PR TITLE
Refactor SSRF protection

### PR DESCRIPTION
**Pull request by mogul**
We think some of the existing BATS tests could move into PHPUnit tests, but we didn't go that far in case there was other value in testing from the "outside".

### CODE REVIEW

File: application/helpers/api_helper.php [hunk 1]
```diff
@@ -305,57 +305,38 @@ function filter_json( $source_datajson, $dataset_array = false ) {
...
-                }
-
-            }
-        }
-
+    // We don't accept unresolvable hostnames or raw IP addresses
+    $host = parse_url($url, PHP_URL_HOST);
+    $resolved = dns_get_record($host, DNS_A + DNS_AAAA);
+    if (!$resolved)
+        return null;
```
Comment (adborden) [inline application/helpers/api_helper.php:325]: Please use `{ }`, even for one-line `if` statements.

File: application/models/Campaign_model.php [hunk 2]
```diff
@@ -1381,9 +1381,7 @@ public function archive_file($filetype, $office_id, $url)
 
         $context = stream_context_create($opts);
 
-        $copy = @fopen($url, 'rb', false, $context);
-        $paste = @fopen($filepath, 'wb');
-
+        $copy = @fopen(filter_remote_url($url), 'rb', false, $context);
```
Comment (adborden) [inline application/models/Campaign_model.php:1384]: This is probably safe, but I think the `archive_file` is dead code that @pburkholder identified. I suspect the URL is being filtered earlier.

File: application/helpers/api_helper.php [hunk 3]
```diff
@@ -305,57 +305,38 @@ function filter_json( $source_datajson, $dataset_array = false ) {
...
     }
     $url = filter_var($url, FILTER_VALIDATE_URL, FILTER_FLAG_PATH_REQUIRED);
+
     // ban non http/https:
+    $allowed_schemes = array('http', 'https');
     $scheme = parse_url($url, PHP_URL_SCHEME);
-    if ($url && !in_array($scheme, $allowed_schemes)) {
-        $url = false;
+    if (!in_array($scheme, $allowed_schemes)) {
+        return null;
```
Comment (adborden) [inline application/helpers/api_helper.php:318]: At one point the distinction between false and null was important, but maybe it was to differentiate an invalid URL from an unresolvable URL but I think https://github.com/GSA/project-open-data-dashboard/commit/32f73be10bc23c0f4ad5b9ed327b1d8b3b14ee00 eliminated the need for that distinction.

You'll want to update this line to accept falsy. https://github.com/GSA/project-open-data-dashboard/blob/e706346f1aca52fd2eda9c42e99ffadfe6881ba4/application/models/Campaign_model.php#L891

File: application/models/Campaign_model.php (hunk 2 above)
Comment (mogul) [reply application/models/Campaign_model.php:1384]: From the README it seems like it's only dead under certain conditions related to S3. We added a test for it recently based on it causing problems during crawls, so it's definitely being invoked.

File: application/helpers/api_helper.php (hunk 3 above)
Comment (mogul) [reply application/helpers/api_helper.php:318]: You're right, and we're returning null everywhere in the function. Will change this.

**Comment by mogul** [comment]
@adborden take another look...?

**Comment by mogul** [comment]
:doh: integration test failing... 👀 now.

**Comment by mogul** [comment]
There was a verrrry subtle bug, in that some dodgy app code seems to expect that the filter will distinguish between "the URL was bad" (false) and "there was no URL" (null). I added a test case for that behavior and made the refactored version match.

**Comment by mogul** [comment]
And by the way, it was good that the integration tests were there to catch this, or it would have broken in prod!

**Pull request merged**